clib/feldspar_array.h

```diff
@@ -36,150 +36,86 @@
 //#define LOG
 #include "log.h"
 
+/* This library contains operations on flat arrays, arrays that do not contain
+   (pointers to) other arrays. For non-flat arrays, these array operations are
+   implemented by code generated from the program by the ArrayOps module.
+   The size argument is always the size in bytes of each element.
+*/
+
 /* TODO qualify the names to avoid clashes with Haskell names */
 
 struct array
 {
     void*    buffer;   /* pointer to the buffer of elements */
     int32_t  length;   /* number of elements in the array */
-    int32_t  elemSize; /* size of elements in bytes; (- sizeof(struct array)) for nested arrays */
-    uint32_t bytes;    /* The number of bytes the buffer can hold */
 };
 
 /* Indexing into an array: */
 /* Result: element of type 'type' */
 #define at(type,arr,idx) (((type*)((arr)->buffer))[idx])
 
-/* Array (re)initialization */
-static inline struct array *initArray(struct array *arr, int32_t size, int32_t len)
+/* Allocate and initialize a struct array if we did not have one already. */
+static inline struct array* allocArray( struct array* src )
 {
-    int newBytes;
+  if( src == NULL ) {
+    src = (struct array *) malloc( sizeof(struct array) );
```

Comment:
Casting prevents the compiler from detecting type issues.

Reply:
I don't follow; this is the standard pattern when using malloc(); type checking will fail without the cast?